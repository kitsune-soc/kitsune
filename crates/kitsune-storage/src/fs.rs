//!
//! File system backed implementation of the [`StorageBackend`] trait
//!

use crate::{Result, StorageBackend};
use async_trait::async_trait;
use bytes::Bytes;
use futures_util::{pin_mut, stream::BoxStream, Stream, StreamExt, TryStreamExt};
use std::path::PathBuf;
use tokio::{
    fs::{self, File},
    io::AsyncWriteExt,
};
use tokio_util::io::ReaderStream;

#[derive(Clone)]
/// File system storage
pub struct Storage {
    storage_dir: PathBuf,
}

impl Storage {
    /// Create a new file system storage
    ///
    /// It always requires a directory the operations on it are relative to
    #[must_use]
    pub fn new(storage_dir: PathBuf) -> Self {
        Self { storage_dir }
    }
}

#[async_trait]
impl StorageBackend for Storage {
    async fn delete(&self, path: &str) -> Result<()> {
        fs::remove_file(self.storage_dir.join(path)).await?;
        Ok(())
    }

    async fn get(&self, path: &str) -> Result<BoxStream<'static, Result<Bytes>>> {
        let file = File::open(self.storage_dir.join(path)).await?;
        Ok(ReaderStream::new(file).map_err(Into::into).boxed())
    }

    async fn put<T>(&self, path: &str, input_stream: T) -> Result<()>
    where
        T: Stream<Item = Result<Bytes>> + Send + 'static,
    {
        let mut file = File::create(self.storage_dir.join(path)).await?;
        pin_mut!(input_stream);
        while let Some(chunk) = input_stream.next().await.transpose()? {
            file.write_all(&chunk).await?;
        }
        file.flush().await?;

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use crate::{fs::Storage, StorageBackend};
    use bytes::{BufMut, BytesMut};
    use futures_util::{future, stream, StreamExt, TryStreamExt};
    use std::str;
    use tempfile::TempDir;

    const TEST_TEXT: &str = r#"
    新時代はこの未来だ
    世界中全部 変えてしまえば 変えてしまえば
    ジャマモノ やなもの なんて消して
    この世とメタモルフォーゼしようぜ
    ミュージック キミが起こす マジック
    目を閉じれば未来が開いて
    いつまでも終わりが来ないようにって
    この歌を歌うよ
    Do you wanna play? リアルゲーム ギリギリ
    綱渡りみたいな旋律 認めない戻れない忘れたい
    夢の中に居させて I wanna be free
    見えるよ新時代が 世界の向こうへ
    さあ行くよ new world
    新時代はこの未来だ
    世界中全部 変えてしまえば 変えてしまえば
    果てしない音楽がもっと届くように
    夢は見ないわ キミが話した「ボクを信じて」
    Ooh
    あれこれいらないものは消して
    リアルをカラフルに越えようぜ
    ミュージック 今始まる ライジング
    目をつぶりみんなで逃げようよ
    今よりイイモノを見せてあげるよ
    この歌を歌えば
    Do you wanna play? リアルゲーム ギリギリ
    綱渡りみたいな運命 認めない戻れない忘れたい
    夢の中に居させて I wanna be free
    見えるよ新時代が 世界の向こうへ
    さあ行くよ new world
    信じたいわ この未来を
    世界中全部 変えてしまえば 変えてしまえば
    果てしない音楽がもっと届くように
    夢を見せるよ 夢を見せるよ 新時代だ
    Ooh
    新時代だ"#;

    #[tokio::test]
    async fn basic() {
        let temp_dir = TempDir::new().unwrap();
        let storage = Storage::new(temp_dir.path().to_owned());

        storage
            .put(
                "hello-world",
                stream::once(future::ok(TEST_TEXT.into())).boxed(),
            )
            .await
            .unwrap();

        let data = storage
            .get("hello-world")
            .await
            .unwrap()
            .try_fold(BytesMut::new(), |mut acc, chunk| {
                acc.put(chunk);
                future::ok(acc)
            })
            .await
            .unwrap();
        let data = str::from_utf8(&data).unwrap();

        assert_eq!(TEST_TEXT, data);

        storage.delete("hello-world").await.unwrap();
        assert!(storage.get("hello-world").await.is_err());
    }
}
