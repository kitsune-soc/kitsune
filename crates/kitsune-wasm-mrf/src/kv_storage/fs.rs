use super::{Backend, BoxError, BucketBackend};
use miette::IntoDiagnostic;
use std::path::Path;

pub struct FsBackend {
    inner: sled::Db,
}

impl FsBackend {
    pub fn from_path<P>(path: P) -> miette::Result<Self>
    where
        P: AsRef<Path>,
    {
        Ok(Self {
            inner: sled::open(path).into_diagnostic()?,
        })
    }
}

impl Backend for FsBackend {
    type Bucket = FsBucketBackend;

    async fn open(&self, module_name: &str, name: &str) -> Result<Self::Bucket, BoxError> {
        self.inner
            .open_tree(format!("{module_name}:{name}"))
            .map(|tree| FsBucketBackend { inner: tree })
            .map_err(Into::into)
    }
}

pub struct FsBucketBackend {
    inner: sled::Tree,
}

impl BucketBackend for FsBucketBackend {
    async fn exists(&self, key: &str) -> Result<bool, BoxError> {
        self.inner.contains_key(key).map_err(Into::into)
    }

    async fn delete(&self, key: &str) -> Result<(), BoxError> {
        self.inner.remove(key)?;
        Ok(())
    }

    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, BoxError> {
        self.inner
            .get(key)
            .map(|maybe_val| maybe_val.map(|val| val.to_vec()))
            .map_err(Into::into)
    }

    async fn set(&self, key: &str, value: &[u8]) -> Result<(), BoxError> {
        self.inner.insert(key, value)?;
        Ok(())
    }
}
