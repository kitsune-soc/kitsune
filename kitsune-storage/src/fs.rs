//!
//! File system backed implementation of the [`StorageBackend`] trait
//!

use crate::{Result, StorageBackend};
use async_trait::async_trait;
use bytes::Bytes;
use futures_util::{stream::BoxStream, Stream, StreamExt, TryStreamExt};
use std::path::PathBuf;
use tokio::{
    fs::{self, File},
    io::AsyncWriteExt,
};
use tokio_util::io::ReaderStream;

/// File system storage
pub struct Storage {
    storage_dir: PathBuf,
}

impl Storage {
    /// Create a new file system storage
    ///
    /// It always requires a directory the operations on it are relative to
    #[must_use]
    pub fn new(storage_dir: PathBuf) -> Self {
        Self { storage_dir }
    }
}

#[async_trait]
impl StorageBackend for Storage {
    async fn delete(&self, path: &str) -> Result<()> {
        fs::remove_file(self.storage_dir.join(path)).await?;
        Ok(())
    }

    async fn get<'a>(&'a self, path: &str) -> Result<BoxStream<'a, Result<Bytes>>> {
        let file = File::open(self.storage_dir.join(path)).await?;
        Ok(ReaderStream::new(file).map_err(Into::into).boxed())
    }

    async fn put(
        &self,
        path: &str,
        input_stream: &mut (dyn Stream<Item = Result<Bytes>> + Send + Unpin),
    ) -> Result<()> {
        let mut file = File::create(self.storage_dir.join(path)).await?;
        while let Some(chunk) = input_stream.next().await.transpose()? {
            file.write_all(&chunk).await?;
        }
        Ok(())
    }
}
