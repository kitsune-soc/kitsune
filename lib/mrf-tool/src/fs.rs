use std::{
    collections::HashMap,
    fs::{self, File},
    io,
    ops::{Deref, DerefMut},
    path::{Path, PathBuf},
};

#[inline]
fn file_not_found() -> io::Error {
    io::Error::new(io::ErrorKind::NotFound, "file not found")
}

pub trait Filesystem {
    type File<'a>: io::Write
    where
        Self: 'a;

    fn copy(&mut self, src: &Path, dst: &Path) -> io::Result<()>;
    fn read(&mut self, path: &Path) -> io::Result<Vec<u8>>;

    fn create_or_truncate(&mut self, path: &Path) -> io::Result<Self::File<'_>>;
    fn open_append(&mut self, path: &Path) -> io::Result<Self::File<'_>>;
}

#[derive(Default)]
pub struct DummyFs {
    inner: HashMap<PathBuf, Vec<u8>>,
}

impl Deref for DummyFs {
    type Target = HashMap<PathBuf, Vec<u8>>;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl DerefMut for DummyFs {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

impl Filesystem for DummyFs {
    type File<'a> = &'a mut Vec<u8>;

    #[inline]
    fn copy(&mut self, src: &Path, dst: &Path) -> io::Result<()> {
        let value = self.inner.get(src).ok_or_else(file_not_found)?;
        self.inner.insert(dst.to_path_buf(), value.clone());

        Ok(())
    }

    #[inline]
    fn read(&mut self, path: &Path) -> io::Result<Vec<u8>> {
        self.inner.get(path).cloned().ok_or_else(file_not_found)
    }

    #[inline]
    fn create_or_truncate(&mut self, path: &Path) -> io::Result<Self::File<'_>> {
        Ok(self.inner.entry(path.to_path_buf()).or_default())
    }

    #[inline]
    fn open_append(&mut self, path: &Path) -> io::Result<Self::File<'_>> {
        self.inner.get_mut(path).ok_or_else(file_not_found)
    }
}

#[derive(Default)]
pub struct NativeFs {
    _priv: (),
}

impl Filesystem for NativeFs {
    type File<'a> = File;

    #[inline]
    fn copy(&mut self, src: &Path, dst: &Path) -> io::Result<()> {
        fs::copy(src, dst)?;
        Ok(())
    }

    #[inline]
    fn read(&mut self, path: &Path) -> io::Result<Vec<u8>> {
        fs::read(path)
    }

    #[inline]
    fn create_or_truncate(&mut self, path: &Path) -> io::Result<Self::File<'_>> {
        File::options()
            .create(true)
            .truncate(true)
            .write(true)
            .open(path)
    }

    #[inline]
    fn open_append(&mut self, path: &Path) -> io::Result<Self::File<'_>> {
        File::options().append(true).open(path)
    }
}
