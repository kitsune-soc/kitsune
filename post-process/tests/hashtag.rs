use pest::Parser;
use post_process::{PostParser, Rule};
use std::fs;

#[test]
fn invalid_hashtag() {
    insta::glob!("input/hashtag/invalid_*", |path| {
        let hashtag = fs::read_to_string(path).unwrap();
        insta::assert_debug_snapshot!(PostParser::parse(Rule::hashtag, &hashtag));
    });
}

#[test]
fn weird_tags() {
    insta::glob!("input/hashtag/weird_*", |path| {
        let post = fs::read_to_string(path).unwrap();
        insta::assert_debug_snapshot!(PostParser::parse(Rule::post, &post).unwrap());
    });
}

#[test]
fn parse_hashtag() {
    insta::glob!("input/hashtag/full_post_*", |path| {
        let post = fs::read_to_string(path).unwrap();
        insta::assert_debug_snapshot!(PostParser::parse(Rule::post, &post).unwrap());
    });
}
