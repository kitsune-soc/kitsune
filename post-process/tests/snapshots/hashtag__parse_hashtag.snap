---
source: post-process/tests/hashtag.rs
expression: "PostParser::parse(Rule::post, &post).unwrap()"
input_file: post-process/tests/input/hashtag/full_post_1
---
[
    Pair {
        rule: text,
        span: Span {
            str: "why am i building a",
            start: 0,
            end: 19,
        },
        inner: [],
    },
    Pair {
        rule: hashtag,
        span: Span {
            str: " #lexer",
            start: 19,
            end: 26,
        },
        inner: [
            Pair {
                rule: component_prefix,
                span: Span {
                    str: " ",
                    start: 19,
                    end: 20,
                },
                inner: [],
            },
            Pair {
                rule: hashtag_content,
                span: Span {
                    str: "lexer",
                    start: 21,
                    end: 26,
                },
                inner: [],
            },
        ],
    },
    Pair {
        rule: text,
        span: Span {
            str: " for",
            start: 26,
            end: 30,
        },
        inner: [],
    },
    Pair {
        rule: hashtag,
        span: Span {
            str: " #posts",
            start: 30,
            end: 37,
        },
        inner: [
            Pair {
                rule: component_prefix,
                span: Span {
                    str: " ",
                    start: 30,
                    end: 31,
                },
                inner: [],
            },
            Pair {
                rule: hashtag_content,
                span: Span {
                    str: "posts",
                    start: 32,
                    end: 37,
                },
                inner: [],
            },
        ],
    },
    Pair {
        rule: text,
        span: Span {
            str: "?",
            start: 37,
            end: 38,
        },
        inner: [],
    },
    Pair {
        rule: hashtag,
        span: Span {
            str: " #龍が如く0",
            start: 38,
            end: 53,
        },
        inner: [
            Pair {
                rule: component_prefix,
                span: Span {
                    str: " ",
                    start: 38,
                    end: 39,
                },
                inner: [],
            },
            Pair {
                rule: hashtag_content,
                span: Span {
                    str: "龍が如く0",
                    start: 40,
                    end: 53,
                },
                inner: [],
            },
        ],
    },
]
